### PR TITLE
feat: support holesky

### LINKED ISSUE #216: add support for `holesky`
https://github.com/eth-clients/holesky

### PR DESCRIPTION
**Pull request by chirag-bgh**
Closes #216

### CODE REVIEW

**Comment by chirag-bgh** [comment]
@ralexstokes where can I find exact values for like `MIN_GENESIS_ACTIVE_VALIDATOR_COUNT` etc?

File: ethereum-consensus/src/clock.rs [hunk 1]
```diff
@@ -12,6 +12,7 @@ use std::{
 pub const MAINNET_GENESIS_TIME: u64 = 1606824023;
 pub const SEPOLIA_GENESIS_TIME: u64 = 1655733600;
 pub const GOERLI_GENESIS_TIME: u64 = 1616508000;
+pub const HOLESKY_GENESIS_TIME: u64 = 1694786100;
```
Comment (ralexstokes) [inline ethereum-consensus/src/clock.rs:15]:
```suggestion
pub const HOLESKY_GENESIS_TIME: u64 = 1694786400;
```

genesis time is `MIN_GENESIS_TIME` and at least `GENESIS_DELAY`

bc of how we are starting the network, we know ahead of time what the genesis time will be

File: ethereum-consensus/src/state_transition/context.rs [hunk 2]
```diff
@@ -262,6 +262,14 @@ impl Context {
         Self::from(phase0_preset, altair_preset, bellatrix_preset, config)
     }
 
+    pub fn for_holesky() -> Self {
+        let config = &configs::holesky::config();
+        let phase0_preset = &phase0::minimal::PRESET;
+        let altair_preset = &altair::minimal::PRESET;
+        let bellatrix_preset = &bellatrix::minimal::PRESET;
```
Comment (ralexstokes) [inline ethereum-consensus/src/state_transition/context.rs:269]: refer: https://github.com/eth-clients/holesky/blob/main/custom_config_data/config.yaml#L2

```suggestion
        let phase0_preset = &phase0::mainnet::PRESET;
        let altair_preset = &altair::mainnet::PRESET;
        let bellatrix_preset = &bellatrix::mainnet::PRESET;
```

File: ethereum-consensus/src/configs/holesky.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    configs::Config,
+    primitives::{Epoch, ExecutionAddress, Gwei, Version, FAR_FUTURE_EPOCH},
+};
+
+// `2**14` (= 16,384)
+pub const MIN_GENESIS_ACTIVE_VALIDATOR_COUNT: usize = 16384;
+pub const MIN_GENESIS_TIME: u64 = 1694786100;
+// 0x00017000
+pub const GENESIS_FORK_VERSION: Version = [0, 1, 112, 0];
+// Genesis delay 5 seconds
```
Comment (ralexstokes) [inline ethereum-consensus/src/configs/holesky.rs:11]:
```suggestion
// Genesis delay 5 minutes
```

File: ethereum-consensus/src/configs/holesky.rs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+pub const GENESIS_DELAY: u64 = 300;
+pub const SECONDS_PER_SLOT: u64 = 12;
+pub const SECONDS_PER_ETH1_BLOCK: u64 = 14;
+pub const MIN_VALIDATOR_WITHDRAWABILITY_DELAY: Epoch = 256;
+pub const SHARD_COMMITTEE_PERIOD: Epoch = 256;
+pub const ETH1_FOLLOW_DISTANCE: u64 = 2048;
+pub const EJECTION_BALANCE: Gwei = 16 * 10u64.pow(9);
```
Comment (ralexstokes) [inline ethereum-consensus/src/configs/holesky.rs:18]: https://github.com/eth-clients/holesky/blob/main/custom_config_data/config.yaml#L61

```suggestion
pub const EJECTION_BALANCE: Gwei = 28 * 10u64.pow(9);
```

File: ethereum-consensus/src/configs/holesky.rs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+pub const ALTAIR_FORK_VERSION: Version = [16, 1, 112, 0];
+pub const ALTAIR_FORK_EPOCH: Epoch = 0;
+// 0x20017000
+pub const BELLATRIX_FORK_VERSION: Version = [32, 1, 112, 0];
+pub const BELLATRIX_FORK_EPOCH: Epoch = 0;
+// 0x30017000
+pub const CAPELLA_FORK_VERSION: Version = [48, 1, 112, 0];
+pub const CAPELLA_FORK_EPOCH: Epoch = 256;
+// 0x40017000
+pub const DENEB_FORK_VERSION: Version = [4, 0, 16, 32];
```
Comment (ralexstokes) [inline ethereum-consensus/src/configs/holesky.rs:32]: https://github.com/eth-clients/holesky/blob/main/custom_config_data/config.yaml#L37

```suggestion
pub const DENEB_FORK_VERSION: Version = [64, 1, 112, 0];
```

File: ethereum-consensus/src/configs/holesky.rs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+pub const INACTIVITY_SCORE_BIAS: u64 = 4;
+pub const INACTIVITY_SCORE_RECOVERY_RATE: u64 = 16;
+pub const PROPOSER_SCORE_BOOST: u64 = 40;
+pub const DEPOSIT_CHAIN_ID: usize = 17000;
+pub const DEPOSIT_NETWORK_ID: usize = 17000;
+
+// Deneb
+pub const MAX_REQUEST_BLOCKS_DENEB: u64 = 128;
+pub const MAX_BLOBS_PER_BLOCK: u64 = 6;
+// MAX_REQUEST_BLOCKS_DENEB*MAX_BLOBS_PER_BLOCK
```
Comment (ralexstokes) [inline ethereum-consensus/src/configs/holesky.rs:44]:
```suggestion
// MAX_REQUEST_BLOCKS_DENEB * MAX_BLOBS_PER_BLOCK
```

File: ethereum-consensus/src/configs/holesky.rs [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+pub const MAX_REQUEST_BLOB_SIDECARS: u64 = 768;
+// 2**12 (= 4096 epochs, ~18 days)
+pub const MIN_EPOCHS_FOR_BLOB_SIDECARS_REQUEST: u64 = 4096;
+pub const BLOB_SIDECAR_SUBNET_COUNT: u64 = 6;
+
+pub fn config() -> Config {
+    // 0
```
Comment (ralexstokes) [inline ethereum-consensus/src/configs/holesky.rs:51]:
```suggestion
```

File: ethereum-consensus/src/configs/holesky.rs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+pub const DEPOSIT_NETWORK_ID: usize = 17000;
+
+// Deneb
+pub const MAX_REQUEST_BLOCKS_DENEB: u64 = 128;
+pub const MAX_BLOBS_PER_BLOCK: u64 = 6;
+// MAX_REQUEST_BLOCKS_DENEB*MAX_BLOBS_PER_BLOCK
+pub const MAX_REQUEST_BLOB_SIDECARS: u64 = 768;
+// 2**12 (= 4096 epochs, ~18 days)
+pub const MIN_EPOCHS_FOR_BLOB_SIDECARS_REQUEST: u64 = 4096;
+pub const BLOB_SIDECAR_SUBNET_COUNT: u64 = 6;
```
Comment (ralexstokes) [inline ethereum-consensus/src/configs/holesky.rs:48]: if we don't need these right here, I'd just leave them out of this PR

if anything they would end up w/ the networking code

File: ethereum-consensus/src/configs/holesky.rs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+pub fn config() -> Config {
+    // 0
+    let terminal_total_difficulty = 0u64.into();
```
Comment (ralexstokes) [inline ethereum-consensus/src/configs/holesky.rs:52]: this works but I think this is a bit more expressive

```suggestion
    let terminal_total_difficulty = U256::zero();
```

**Review by ralexstokes — CHANGES_REQUESTED** [review]
thanks for this!

left a few things to fix but then this is ready to merge

File: ethereum-consensus/src/clock.rs (hunk 1 above)
Comment (ralexstokes) [reply ethereum-consensus/src/clock.rs:15]: can you please update the constant?

File: ethereum-consensus/src/configs/holesky.rs [hunk 10]
```diff
@@ -0,0 +1,88 @@
+use ssz_rs::U256;
+
+use crate::{
+    configs::Config,
+    primitives::{Epoch, ExecutionAddress, Gwei, Version, FAR_FUTURE_EPOCH},
+};
```
Comment (ralexstokes) [inline ethereum-consensus/src/configs/holesky.rs:6]: let's keep imports all together

```suggestion
use ssz_rs::U256;
use crate::{
    configs::Config,
    primitives::{Epoch, ExecutionAddress, Gwei, Version, FAR_FUTURE_EPOCH},
};
```

**Review by ralexstokes — CHANGES_REQUESTED** [review]
looks good! one nitpick on the imports formatting

**Review by ralexstokes — CHANGES_REQUESTED** [review]
one more thing on import

File: ethereum-consensus/src/configs/holesky.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (chirag-bgh) [reply ethereum-consensus/src/configs/holesky.rs]: Done!

**Review by ralexstokes — APPROVED** [review]

**Pull request merged**